### PR TITLE
Add provider name to provider resource

### PR DESCRIPTION
**Pull request by jplewa**
## Task

Resolves: #228 

## Description

Disclaimer: for providers with complex names (such as `aws-native`), the names have the following format: `AzurenativeProviderResource`. I'll try to figure out if there's an easy way to make the name `AzureNativeProviderResource` instead, but the easiest way was to ignore the capitalization of complex names.

### TODO

- [x] ~~Add tests~~ Fix existing tests
- [x] If the current solution is determined to be not good enough, figure out a way to capitalize complex names

### CODE REVIEW

File: src/main/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiName.kt [hunk 1]
```diff
@@ -16,9 +16,21 @@ import org.virtuslab.pulumikotlin.codegen.utils.decapitalize
...
+    val moduleName: String?,
     val name: String,
+    val isProvider: Boolean,
 ) {
+    val namespace: List<String>
+        get() = (
+            baseNamespace +
+                (providerNameOverride ?: providerName) +
+                moduleName
+            )
```
Comment (myhau) [inline src/main/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiName.kt:30]: It's a bit hard to read. 

One way of improving it:

```kt
val effectiveProviderName =  (providerNameOverride ?: providerName)
val namespaceSections = baseNamespace + effectiveProviderName + moduleName
namespaceSections
   .filter...
   ....
```

File: src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt [hunk 2]
```diff
@@ -27,25 +28,28 @@ internal class PulumiNameTest {
         TypeName(
             "aws:acm/CertificateOptions:CertificateOptions",
             namingConfigurationWithSlashInModuleFormat("aws"),
-            PulumiName("aws", listOf("com", "pulumi", "aws", "acm"), "CertificateOptions"),
+            PulumiName("aws", null, listOf("com", "pulumi"), "acm", "CertificateOptions", false),
```
Comment (myhau) [inline src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt:31]: (minor)

Would it be possible to get rid of this `null` and `false`? 

Another `PulumiName` factory method might be the solution here?

File: src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt [hunk 3]
```diff
@@ -304,8 +398,13 @@ internal class PulumiNameTest {
...
-            { assertEquals("overrideProvider", pulumiName.providerName) },
-            { assertEquals(listOf("org", "example", "overrideProvider", "overrideModule"), pulumiName.namespace) },
+            { assertEquals("provider", pulumiName.providerName) },
+            {
+                assertEquals(
+                    listOf("org", "example", "overrideProvider", "overrideModule"),
+                    pulumiName.namespace,
+                )
+            },
             { assertEquals("ObjectName", pulumiName.name) },
```
Comment (myhau) [inline src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt:408]: (not really about this PR and nothing we can do now?)

I'm not sure if we should have such tests. We should test the public interface (e.g. `.toPackage`, `toFunctionGroupObjectName`) and I'm not sure `providerName`, `namespace` and `name` should be public fields.

**Review by myhau — APPROVED** [review]
LGTM

I wish I could delve into this more, but two things that I noticed (partially related to this PR)

- `isProvider` — passing this primitive argument through all the layers feels wrong to me. Maybe we should extend `UsageKind` by adding another `Subject` (`Provider`) or take a step back and think of a better design altogether? 
- `PulumiName` —  I think we could refactor it and its usage a bit along with `NamingConfiguration` + `UsageKind` abstractions

**Comment by jplewa** [comment]
> I wish I could delve into this more, but two things that I noticed (partially related to this PR)
> 
>- `isProvider` — passing this primitive argument through all the layers feels wrong to me. Maybe we should extend `UsageKind` by adding another `Subject` (`Provider`) or take a step back and think of a better design altogether?
> 
>- `PulumiName` —  I think we could refactor it and its usage a bit along with `NamingConfiguration` + `UsageKind` abstractions

Both good points. I tried to address the first one, but adding this new enum value isn't as straightforward as it seems (`Subject` isn't passed to `PulumiName` and I think we'd need to add an extra field in `Modifiers` which seems a bit clumsy). I'll create a task to address both of these things, but for now I'll merge it as is.

File: src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt (hunk 3 above)
Comment (jplewa) [reply src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt:408]: True, I'll create a task 👍

Comment (jplewa) [reply src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt:408]: https://github.com/VirtuslabRnD/pulumi-kotlin/issues/273

**Pull request merged**